### PR TITLE
Cookie whitelist

### PR DESCRIPTION
**Pull request by jeroen7s**
PR for adding ability to whitelist cookies #30
i'm not a JS developer nor a chrome extension dev
so feedback is definitely welcome

~TODO: fix button styling~
~TODO: change button icon if whitelisted~
~TODO: remove item from whitelist if clicked on a button of a whitelisted cookie~
~TODO: don't add duplicate whitelist items~
~TODO: make whitelist a dictionary instead of an array~
~TODO: rewriting logic to detect if cookie is in whitelist~
~TODO: add button to whitelist domains instead of specific cookies~
~TODO: reusing logic for delete all cookies button~

### CODE REVIEW

**Comment by Manvel** [comment]
Thanks @jeroen7s for the PR I won't have a chance to review the PR earlier than weekend, but as I see there are still a lot of TODOs, I assume the PR is not yet ready for the review, right ?

**Comment by Manvel** [comment]
> i'm not a JS developer nor a chrome extension dev
> so feedback is definitely welcome

I don't have yet any linter, or contribution guide setup yet and probably there is a lot of refactoring I need to do one day, so I'll really appreciate if you keep the coding/naming consistent with current implementation. The more readable is code and consistent the easier for me would be to review it.

File: js/back.js [hunk 1]
```diff
@@ -65,6 +65,38 @@
...
         return accumulator;
       }, {});
+      if (browsingDataObj.cookies) {
+        // delete cookies here + ignore whitelisted cookies
+        getStorage("cookieWhitelist", function(data) {
+          var domain_list = data.cookieWhitelist
+          getAllCookies({}, function(cookies)
+          {
+            var callbackCount = 0;
+            for (var i = 0; i < cookies.length; i++)
```
Comment (Manvel) [inline js/back.js]: Suggestion: if we don't use index, we can use `for..of` loop instead, in most of the cases `for` loop were used as initial code was written even before `for..of` were introduced. This way the code will be more readable.

```
for (const cookie of cookies)
{
  ...
}
```

File: js/back.js [hunk 2]
```diff
@@ -65,6 +65,38 @@
...
+            {
+              var cookie = cookies[i];
+              var url = getUrl(cookie.domain, cookie.path, cookie.secure);
+
+              var wl = false;
```
Comment (Manvel) [inline js/back.js]: Suggestion: please use `whitelist` as the variable name, while it's obvious for me that `wl` stands for whitelist, for another person who reads the code, or if I decide to read it after couple of months, I'll need to remember what was this all about.

File: js/back.js [hunk 3]
```diff
@@ -65,6 +65,38 @@
...
+              for (var x in domain_list) { 
+                var wl_domain = domain_list[x][0]
```
Comment (Manvel) [inline js/back.js]: Nit picking: I have doing nit picking, but I have to as I don't have linters setup yet(my bad ). May I ask you to use camelCase for variables, as it's done in other places as well.

**Comment by jeroen7s** [comment]
small question
is it possible to add an additional parameter (like name, storeId, httpOnly, secure etc...) called whitelist to a cookie, or would that create a security issue?
not sure if sites would be able to get or alter that parameter if added
then we don't need a separate array of arrays (or dictionary) to store the whitelist,

**Comment by jeroen7s** [comment]
welp, will not have much time this weekend after all, guess it will take a bit longer

**Comment by Manvel** [comment]
@jeroen7s No worries - take your time, thanks a lot for providing PR though.

Currently I'm busy with another project so I'll not annoy you with the Merge conflicts soon :)

**Comment by jeroen7s** [comment]
I think most of the work is done now
I only need to figure out how to let the UI know which cookies/domains have been white-listed
and replace the current ugly broken icon with a better one, with an checked/unchecked state icon
but that will be for this weekend I think
or maybe tomorrow or friday-evening
you should be able to review most of the code already :)

**Comment by Manvel** [comment]
> you should be able to review most of the code already :) 

I'll try to have a skim before the weekend, otherwise I'll do full review, on weekend.

File: js/ui/tab_cookies.js [hunk 4]
```diff
@@ -164,7 +200,7 @@
         break;
       case "delete-domain-cookies":
         var domain = getParentData(element, "data-access");
-        getAllCookies({"domain": domain}, function(cookies)
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: Why was domain cookies removal functionality removed ? This shouldn't be necessary.
This is triggered when user hit delete key when focused on a top level list item(Domain item) or when "x" button next to the list item is clicked, see -> https://github.com/Manvel/Privacy-Manager/blob/master/popup.html#L113

File: js/ui/tab_cookies.js [hunk 5]
```diff
@@ -156,6 +156,52 @@
           }
         });
         break;
+      case "whitelist-cookie-domain":
+        var domain = getParentData(element, "data-access");
+        var url = getUrl(domain, "", true);
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: Question: You get/generate here URL, but it's not being used, same applies below.

In general I'm not sure why you need to determine the actual URL:
- Currently PM is doesn't differentiate between http and https when it comes to domain.

I think if you wish to whitelist domain, you need only domain rather than differentiate between protocols.
Protocols consideration might be a separate issue, if we decide to differentiate between them.

With above said, I think you should be able to achieve your result without implementing `getUrl ` method for this usecase.

File: js/ui/tab_cookies.js [hunk 6]
```diff
@@ -156,6 +156,52 @@
...
+        getStorage("cookieWhitelist", function(data) 
+        {
+          let whitelist = data["cookieWhitelist"]
+          if (!whitelist.propertyIsEnumerable(domain))
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: Detail: I assume you wish to check here if whitelist has "domain" property.

I would suggest to use:
```
if ("domain" in whitelist)
```

This will make code more readable and intention more clear.

File: js/ui/tab_cookies.js [hunk 7]
```diff
@@ -156,6 +156,52 @@
...
+        var domain = getParentData(element, "data-access");
+        var url = getUrl(domain, "", true);
+        getStorage("cookieWhitelist", function(data) 
+        {
+          let whitelist = data["cookieWhitelist"]
+          if (!whitelist.propertyIsEnumerable(domain)) 
+          {
+            whitelist[domain] = []
+          }
+          if (!whitelist[domain].includes(""))
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: Question: Not sure what is the intention here of checking for an empty item in the array ? 

I assume what you want to do here is to check whether user wish to whitelist whole domain or a specific cookie for a domain:

Suggestion: For that reason I would think of a current data structure for storing "cookieWhitelist":
```
{
  cookieWhitelist: {
    "amazon.com": 
    {
      domainWhitelist: false,
      cookies: ["aws-userInfo", "session"...]
    },
    "login.google.com":
    {
      domainWhitelist: true,
      cookies: []
    }
  }
  ...
} 
```
```
const domain = getParentData(element, "data-access");

getStorage("cookieWhitelist", function(cookieWhitelist)
{
    if (domain in cookieWhitelist && cookieWhitelist[domain].domainWhitelist)
    {
      cookieWhitelist[domain].domainWhitelist = false;
    }
    else
    {
      // set domainWhitelist = true;
    }
}
```

This way it would be possible to more easily manage the whitelisted cookies and domains and not loose the states when switching between whitelisting enabled and disabled states.

(Unless I'm missing something)

Comment (Manvel) [reply js/ui/tab_cookies.js]: Removing and adding items into cookies array for specific domain I think should be handled in `whitelist-sublist-cookie`.

File: js/ui/tab_cookies.js (hunk 5 above)
Comment (jeroen7s) [reply js/ui/tab_cookies.js]: yup, was not really needed, so i removed it

File: js/ui/tab_cookies.js (hunk 7 above)
Comment (jeroen7s) [reply js/ui/tab_cookies.js]: > Question: Not sure what is the intention here of checking for an empty item in the array ?
yup, this is correct, as cookie names can't be empty, it's easy to check against an empty string as the "domain"
though that's not clear when reading the code
the suggested structure is definitely better

Comment (jeroen7s) [reply js/ui/tab_cookies.js]: > I assume what you want to do here is to check whether user wish to whitelist whole domain or a specific cookie for a domain:

not completely correct
it's simply to check if a domain was already whitelisted
since an empty string is stored if a domain is whitelisted (since cookie names can't be empty, it's easy to store just an empty string as a domain whitelist)
this is not related to whitelisting specific cookies for a domain
though it could be more clear

the logic for whitelisting specific cookies is in whitelist-sublist-cookie

Comment (Manvel) [reply js/ui/tab_cookies.js]: > I assume what you want to do here is to check whether user wish to whitelist whole domain or a specific cookie for a domain:
>  > it's simply to check if a domain was already whitelisted

I forgot to edit my message, after finalizing the suggestion. You are correct, this part of the code is only related to Domain whitelist.

> since cookie names can't be empty, it's easy to store just an empty string as a domain whitelist

I see.

> though it could be more clear

I think if you would keep something similar to a structure mentioned above, it would be more readable and intuitive.

File: js/ui/tab_cookies.js (hunk 5 above)
Comment (Manvel) [reply js/ui/tab_cookies.js]: I still can see it, maybe you forgot to push your local changes yet?

File: js/ui/tab_cookies.js [hunk 8]
```diff
@@ -156,6 +156,52 @@
...
+        getStorage("cookieWhitelist", function(data) 
+        {
+          let whitelist = data["cookieWhitelist"]
+          if (!whitelist.propertyIsEnumerable(domain)) 
+          {
+            whitelist[domain] = []
+          }
+          if (!whitelist[domain].includes(accessObj.cookie)) 
+          {
+            data["cookieWhitelist"][domain].push(accessObj.cookie)
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: Detail: You can reference here to `whitelist `, same as you have done below:
```
whitelist[domain]...
```

Also `setStorage()` can be moved out of the `if-else` cases to avoid duplications:

**Comment by jeroen7s** [comment]
updated the code now, implemented the feedback
now the only thing that remains is updating the UI and changing the icons
however the logic for that seems a bit complicated atm (at least from a non-js dev's perspective)
and i'm not sure how I should go about it
maybe you could take a look at implementing the UI bits yourself or point me in the right direction?

**Comment by Manvel** [comment]
> updated the code now, implemented the feedback

I'll again try to review that during the week, otherwise will do complete review over the weekend.

> however the logic for that seems a bit complicated atm (at least from a non-JS / non-UI dev's perspective)

I see, yes it might seem to be complicated, here are some thoughts, hope they will make the task simple:
- Differentiate the states
- Reflect list item whitelist state when loaded
- Detect local storage change
- Reflect change on the UI accordingly

# Differentiate the states
Currently the differentiation is done using [data attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) for example `data-expanded="true"` , maybe you can use something like `data-whitelisted="true"` and create CSS accordingly, smth like:
```
.tableList > li[data-whitelisted="true"] > div .icon.whitelist
{
  background-image: url(...);
}

.tableList li > ul li[data-whitelisted="true"] .icon.whitelist
{
  background-image: url(...);
}
```

# Reflect list item whitelist state when loaded

We have two places where we are loading cookies, first when we are loading [Domain cookies](https://github.com/Manvel/Privacy-Manager/blob/master/js/ui/tab_cookies.js#L91) and when we are loading [Cookies for domain](https://github.com/Manvel/Privacy-Manager/blob/master/js/ui/tab_cookies.js#L146).

```
if (whitelisted)
  list.dataset.whitelisted = "true";
```

# Detect local storage change
Chrome storage API does have [onchange event](https://developer.chrome.com/extensions/storage#event-onChanged), this can be used for detecting whitelisting changes and update the table accordingly.

We are already detecting cookies change and [act accordingly here](https://github.com/Manvel/Privacy-Manager/blob/master/js/ui/tab_cookies.js#L27). So we might need something like `onWhitelistChange` and act accordingly when the state in the local storage is changed.

# Reflect change on the UI accordingly
Reflecting changes on the UI I think can be done similarly as in the `onCookiesChange`, there is [updateItem](https://github.com/Manvel/Privacy-Manager/blob/master/js/ui/common.js#L778) method which accept modified element and accessor, this way should be possible to update the Cookie list state.

```
const accessor = domain;
domainElement.dataset.whitelisted = state;
tableList.updateItem(domainElement, accessor);

// Similar way is done for sublist
```

Hope this helps.

**Comment by Manvel** [comment]
> though there doesn't seem to be a
> `tableList.updateItem(domainElement, accessor);`
> equivalent for sublist items

Hmm, that's true :(

If you notice that it's not trivial to do that, I'll also be completely fine with a temporarily solution if there would be a separate ticket describing that(I can create one), to keep the review simple enough.

By temporarily solution I mean, smth along the lines:

```
// js/ui/tab_cookies.js
...
function onWhitelistChange(...)
{
  ...
  if (subItem)
  {
     querySelector([`data-access=${accessor}`]).dataset.whitelisted = whitelistStatus;
  }
}
```

The future plan is to make a web component out of the table list(Also from other components), implement proper tests, to make new features introduction easy and safe.

**Comment by jeroen7s** [comment]
thanks, I updated the code, with the workaround.
also added code to update the whitelist status on loading the cookie tab and when loading domain cookies.
so everything is basically functional, though it can still be improved.
the UI styling needs some work
as it currently looks like this:
image

File: css/main.css [hunk 9]
```diff
@@ -527,6 +531,52 @@ button.icon.edit:after
   background-image: url(/css/img/edit.svg);
 }
 
+/* button[role="checkbox"][aria-checked="true"] > span:first-child
```
[the remark this answers is not in the document]
Comment (Manvel) [reply css/main.css]: Mixed rules: This rules apply to the 

I assume that's why the UI [you mentioned](https://github.com/Manvel/Privacy-Manager/pull/31#issuecomment-442936125) is broken.

I think if you remove the label markup:
```
<span id="btn-on-label">wl</span>
<span id="btn-off-label">bl</span>
```
The UI will stop being broken as the Table list component doesn't expect "toggle switch" to be inside of items.

I assume what you want to do is to use "::after" rule to be applied to apply the icon to whitelisting buttons, something like this(similar to the "delete button"):

```
.tableList li > ul li[data-whitelist="true"] .icon.whitelist::after
{
  background-image: url(/css/img/check-mark-filled.svg);
}
```

File: css/main.css [hunk 10]
```diff
@@ -527,6 +531,52 @@ button.icon.edit:after
...
+  color: #fff;
+}
+
+button[role="checkbox"][aria-checked="true"] > span:last-child
+{
+  color: #000;
+  background: none;
+} */
+
+.tableList > li[data-whitelist="true"] > div .icon.whitelist
```
Comment (Manvel) [inline css/main.css]: Following the comment above I think those rule shouldn't be needed as well.

File: js/ui/tab_cookies.js [hunk 11]
```diff
@@ -371,19 +419,54 @@
...
       }
     };
   }
-
+  function updateWhitelistInList(change)
+  {
+    if ("cookieWhitelist" in change) 
+    {
+      let newValue = change.cookieWhitelist.newValue;
+      for (var domain in newValue)
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: Suggestion: I can imagine that iterating over whole table and updating items using Browser API can have performance implications, what we can do here is to compare old and new values and act accordingly instead of querying each item.

Let me know please if you want me to elaborate on this with code examples.

File: css/main.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeroen7s) [reply css/main.css]: changed it, now seems to work and look correct 👍

[the remark this answers is not in the document]
Comment (jeroen7s) [reply css/main.css]: also the commented out part was copy-pasted before, to use as example for the whitelist buttons css
so I deleted it

File: js/ui/tab_cookies.js (hunk 11 above)
Comment (Manvel) [reply js/ui/tab_cookies.js]: There is two solutions I see.
Determine difference and act:
```
onStorageChange.addListener(function(changeInfo)
{
  if ("cookieWhitelist" in change)
  {
    let {newValue, oldValue} = change.cookieWhitelist;
    let domains = compare(newValue, oldValue);
    updateWhitelistingList(domains);
  }
});
```

Comparing when requesting items from memory `tableList.getItem(domain);`:

```
let domainItem = tableList.getItem(domain);

if (domainItem && domainItem. dataset.whitelist == newValue[domain].domainWhitelisted)
{
...
}
```

Comment (jeroen7s) [reply js/ui/tab_cookies.js]: added some code to filter out unchanged domains (domainWhitelist and cookies both unchanged) before calling updateWhitelistInList
also updated getcookies to only send current domain to updateWhitelistInList
so the only time updateWhitelistInList should go through everything is if the cookies tab is filled for the first time

File: js/ui/tab_cookies.js [hunk 12]
```diff
@@ -371,19 +423,89 @@
...
+  onStorageChange.addListener(function(changeInfo)
+  {
+    if ("cookieWhitelist" in changeInfo)
+    {
+      let newValue = changeInfo.cookieWhitelist.newValue
+      let oldValue = changeInfo.cookieWhitelist.oldValue
+      let changed = {}
+      if (oldValue) 
+      {
+        for (let key in newValue)
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: I wonder if something like this would be enough:

```
const diff = Object.keys(newValue).reduce((acc, domain) => 
  {
    if (JSON.stringify(oldValue[domain]) != JSON.stringify(newValue[domain]))
    {
      acc[domain] = newValue[domain];
    }
    return acc;
  }, {});
```

Considering that out JSON objects are not big and complex, I don't think that Stringify can come too much costy, but if you would have that concern you can use `reduce` and simplify the existing logic, what you think ?

If needed can provide example for current implementation simplification.

Comment (jeroen7s) [reply js/ui/tab_cookies.js]: this definitely seems like an improvement
didn't know about the existence reduce, but seems like a better way
JSON.stringify can definitely reduce the complexity of the code a lot
the only downside i see is that if somehow the order of the cookies in oldValue[domain].cookies changes it would consider it a changed object, while the current implementation would consider them the same, don't think that would have any real impact tho
it would definitely be an improvement overall, will check it out and change it later

Comment (Manvel) [reply js/ui/tab_cookies.js]: You are right, even though in most of the cases it would work, but mostly because how v8 might deal with the simple Objects and order of the keys in the object, but that should work more reliable for the Arrays as we only have String values, so maybe something like this can be even better solution:

```
const newDomainObj = newValue[domain];
const oldDomainObj = oldValue[domain];

if (newDomainObj. domainWhitelist != oldDomainObj.domainWhitelist ||
    JSON.stringify(newDomainObj.cookies) != JSON. stringify(oldDomainObj.cookies))
{
 ... 
}
```

File: popup.html [hunk 13]
```diff
@@ -116,6 +116,7 @@ <h4 data-i18n="cookies_general"></h4>
 									<span data-text="domain" class="domainName"></span>
 									<span data-text="cookienum" class="cookiesNumber"></span>
 								</div>
+								<button role="checkbox" data-action="whitelist-cookie-domain" data-i18n="whitelistCookieDomain" data-info="true" class="icon whitelist"></button>
```
Comment (Manvel) [inline popup.html]: Regarding `data-i18n`: Currently there is no easy way to set the title of the `title` attribute of the element using HTML.

`data-i18n` is being used to populate the content of the item it's set on, see -> https://github.com/Manvel/Privacy-Manager/blob/master/js/ui/common.js#L173 . But even in this case the content of `<button>` will not be populated because we load items dynamically and call i18n function on `DOMContentLoaded ` event which happens before we populate the items.

I think what you want(ed) to do here is to add the `whitelistCookieDomain` string during [table list item load](https://github.com/Manvel/Privacy-Manager/blob/master/js/ui/common.js#L533).

I would imagine that we might want to have is a `[data-i18n-title]` attribute which should handle the title population, but for dynamic items we might want to setup a cache.

We can either tackle that in current review or separately.

File: css/main.css [hunk 14]
```diff
@@ -527,6 +531,34 @@ button.icon.edit:after
   background-image: url(/css/img/edit.svg);
 }
 
+.tableList > li > div .icon.whitelist::after
+{
+  background-image: url(/css/img/check-mark-circle.svg);
+  background-size: 105%;
```
Comment (Manvel) [inline css/main.css]: Was there a reason you used 105% ?
This causes the icon on my Mac to be cut:
screen shot 2018-12-16 at 5 04 52 pm

File: js/back.js [hunk 15]
```diff
@@ -18,21 +18,27 @@
 
 "use strict";
 
-(function(global)
+(function (global)
```
Comment (Manvel) [inline js/back.js]: Detail: Can you please fix this autocorrections(see also below) for the unrelated code please.

File: css/main.css (hunk 14 above)
Comment (Manvel) [reply css/main.css]: 100% looks fine on my screen.

Comment (jeroen7s) [reply css/main.css]: yeah i wanted to make the icon a little bit bigger, but now you mention it it does look a bit cut
will set it back to 100

File: popup.html (hunk 13 above)
Comment (jeroen7s) [reply popup.html]: yeah, i hoped it would work just like the regular (i) info buttons when you over over it

Comment (jeroen7s) [reply popup.html]: not exactly sure how i should implement this though

Comment (Manvel) [reply popup.html]: > not exactly sure how i should implement this though

## Solution 1
I can see two ways how it's possible to achieve that, there is `_itemFromTmpl` and `_updateListElem` methods, the logic can be implemented in either of those I assume.

I would imagine something like this:

```
const listElement = ...;

for (const i18nTitleElement of listElement.querySelectorAll("[data-i18n-title]"))
  i18nTitleElement.title = getMsg(i18nTitleElement.dataset.i18nTitle);
```

regarding cache: I would assume most of the browsers do tackle the caching, but alternatively that can be implemented in `getMsg` in `utils.js`, maybe smth like this(The logic can probably be simplified):

```
function getMsg(text, cache)
{
  let result = msgsCache[text];
  if (result)
    return result;
  else
    result = chrome.i18n.getMessage(text) || text;
  if (cache && !(text in msgsCache))
    msgsCache[text] = text;
  return result;
}
```

## Solution 2

Maybe we can also do something even better, by passing the actual text to the `itemObj` by introducing new type of data, in order to populate the text Table List is using `data-text` attribute to locate the element and update the text, so we will need to introduce something like this, considering https://github.com/Manvel/Privacy-Manager/blob/master/js/ui/tab_cookies.js#L374 :

```
function createDomainObj(domain, cookienum, whitelist = false)
  {
    return {
      dataset: {
        access: domain,
        whitelist: whitelist
      },
      texts: {
        domain: domain,
        cookienum: cookienum + " Cookies"
      },
     titles: {
        whitelist: whitelistCookieDomain
     }
    };
  }
```

And in the https://github.com/Manvel/Privacy-Manager/blob/master/js/ui/common.js#L699 

```
var titleObjs = itemObj.titles;
for (var title in titleObjs)
{
  var titleElement = listElem.querySelector("[data-text='"+ title +"']");
  if (titleElement)
    textElement.title = titleObjs[title];
}
```

And in this case we don't need to Cache anything and the logic should be consistent.
(The logic also can be simplified)

I like 2-nd solution more than 1-st one.

Comment (jeroen7s) [reply popup.html]: updated code, now message displays correctly when hovering over the whitelist buttons
used the 2nd solution you proposed for this, seems to work well

**Comment by jeroen7s** [comment]
should pretty much be done now
let me know if theres anything left to be done

File: js/ui/tab_cookies.js [hunk 16]
```diff
@@ -367,23 +419,88 @@
       texts: {
         name: cookie.name,
         value: cookie.value
+      },
+      titles: {
+         whitelist: getMsg("whitelistSublistCookie")
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: As mentioned in the solution, I was hoping we can get the message from the cache, rather than relying on the Browser API to cache current request.

File: js/ui/tab_cookies.js [hunk 17]
```diff
@@ -367,23 +419,88 @@
...
       }
     };
   }
 
-  function createDomainObj(domain, cookienum)
+  function createDomainObj(domain, cookienum, whitelist = false)
```
Comment (Manvel) [inline js/ui/tab_cookies.js]: Seems like we are not using `whitelist = false` parameter now, but I don't see harm of keeping it here, though a comment would have been useful.

File: js/ui/tab_cookies.js (hunk 16 above)
Comment (Manvel) [reply js/ui/tab_cookies.js]: So basically we just need to declare `getMsg("whitelistSublistCookie")` outside of current function. Having that in top of the file should be enough smth like:

```
const cookieWhitelistButtonTitle = getMsg("whitelistSublistCookie");
const domainWhitelistButtonTitle = ...
```

File: js/ui/tab_cookies.js (hunk 17 above)
Comment (Manvel) [reply js/ui/tab_cookies.js]: Something like one below should be sufficient:
```
  /**
   * Create a Table List Item Structure Object
   * @param {String} domain Domain name
   * @param {Number} cookienum Number of domain cookies
   * @param {Boolean} whitelist specifies whether domain is whitelisted
   */
```

**Comment by Manvel** [comment]
We are almost there.

Besides of 2 comments I have left, there are still some nits that needs to be fixed.

Regarding comments: If you are fine with the comments, can you please add those two fixes as well ?

Regarding Nit picking: I think they shouldn't be discussed in the review, but rather should be test implemented for detecting them. So let's keep them outside of the scope of current PR, I'll followup and fix them separately and setup a Travis CI which will spot inconsistencies automatically, after we can adapt the linters if any disagreement regarding styling.

Basically: Please fix 2 comments when you have time, I'll take care for the rest.

**Comment by Manvel** [comment]
That was fast :) LGTM and thanks a lot for the PR 🎉
